src/test/regress/sql/merge.sql

```diff
@@ -1206,6 +1207,111 @@ SET citus.log_remote_commands to false;
 SELECT compare_tables();
 ROLLBACK;
 
+
+-- let's create source and target table
+CREATE TABLE source_pushdowntest (id integer);
+CREATE TABLE target_pushdowntest (id integer );
+
+-- let's distribute both table on id field
+SELECT create_distributed_table('source_pushdowntest', 'id');
+SELECT create_distributed_table('target_pushdowntest', 'id');
+
+-- we are doing this operation on single node setup let's figure out colocation id of both tables
+-- both has same colocation id so both are colocated.
+select colocationid,logicalrelid from pg_dist_partition where logicalrelid = 'source_pushdowntest'::regclass OR logicalrelid = 'target_pushdowntest'::regclass;
+
+SET client_min_messages TO DEBUG1;
+-- Test 1 : tables are colocated AND query is multisharded AND Join On distributed column : should push down to workers.
+
+EXPLAIN MERGE INTO target_pushdowntest t
+USING source_pushdowntest s
+ON t.id = s.id
+WHEN NOT MATCHED THEN
+  INSERT (id)
+  VALUES (s.id);
+
+-- Test 2 : tables are colocated AND source query is not multisharded : should push down to worker.
+
+EXPLAIN MERGE INTO target_pushdowntest t
+USING (SELECT * from source_pushdowntest where id = 1) s
+on t.id = s.id
+WHEN NOT MATCHED THEN
+  INSERT (id)
+  VALUES (s.id);
+
+
+-- Test 3 : tables are colocated source query is single sharded but not using source distributed column in insertion. let's not pushdown.
+INSERT INTO source_pushdowntest (id) VALUES (3);
+
+EXPLAIN MERGE INTO target_pushdowntest t
+USING (SELECT 1 as somekey, id from source_pushdowntest where id = 1) s
+on t.id = s.somekey
+WHEN NOT MATCHED THEN
+  INSERT (id)
+  VALUES (s.somekey);
+
+
+-- let's verify if we use some other column from source for value of distributed column in target.
+-- it should be inserted to correct shard of target.
+CREATE TABLE source_withdata (id integer, some_number integer);
+CREATE TABLE target_table (id integer, name text);
+SELECT create_distributed_table('source_withdata', 'id');
+SELECT create_distributed_table('target_table', 'id');
+
+INSERT INTO source_withdata (id, some_number) VALUES (1, 3);
+
+-- we will use some_number column from source_withdata to insert into distributed column of target.
+-- value of some_number is 3 let's verify what shard it should go to.
+select worker_hash(3);
+
+-- it should go to second shard of target as target has 4 shard and hash "-28094569" comes in range of second shard.
+MERGE INTO target_table t
+USING (SELECT id, some_number from source_withdata where id = 1) s
```

Comment:
Ideally we should have a "random" insert of integer range and verify that it is getting inserted correctly into the right shard, and be able to select from co-ordinator, to make sure we don't have any phantom rows.